log/loki.go

```diff
@@ -0,0 +1,486 @@
+		switch key {
+		case "pushPeriod":
+			h.pushPeriod, err = time.ParseDuration(value)
+			if err != nil {
+				return nil, fmt.Errorf("couldn't parse the loki pushPeriod %w", err)
+			}
+		case "profile":
+			h.profile = true
+		case "limit":
+			h.limit, err = strconv.Atoi(value)
+			if err != nil {
+				return nil, fmt.Errorf("couldn't parse the loki limit as a number %w", err)
+			}
+		case "msgMaxSize":
+			h.msgMaxSize, err = strconv.Atoi(value)
+			if err != nil {
+				return nil, fmt.Errorf("couldn't parse the loki msgMaxSize as a number %w", err)
+			}
+		case "level":
+			h.levels, err = getLevels(value)
+			if err != nil {
+				return nil, err
+			}
+		default:
+			if strings.HasPrefix(key, "label.") {
+				labelKey := strings.TrimPrefix(key, "label.")
+				h.labels = append(h.labels, [2]string{labelKey, value})
+				continue
+			}
+
+			return nil, fmt.Errorf("unknown loki config key %s", key)
+		}
+	}
+
+	h.client = &http.Client{Timeout: h.pushPeriod}
+
+	go h.loop()
+	return h, nil
+}
+
+func getLevels(level string) ([]logrus.Level, error) {
+	// TODO figure out if `tracing`,`fatal` and `panic` should be included
+	lvl, err := logrus.ParseLevel(level)
+	if err != nil {
+		return nil, fmt.Errorf("unknown log level %s", level) // specifically use a custom error
+	}
+	index := sort.Search(len(logrus.AllLevels), func(i int) bool {
+		return logrus.AllLevels[i] > lvl
+	})
+	return logrus.AllLevels[:index], nil
+}
+
+// fill one of two equally sized slices with entries and then push it while filling the other one
+// TODO benchmark this
+//nolint:funlen
+func (h *lokiHook) loop() {
+	var (
+		msgs       = make([]tmpMsg, h.limit)
+		msgsToPush = make([]tmpMsg, h.limit)
+		dropped    int
+		count      int
+		ticker     = time.NewTicker(h.pushPeriod)
+		pushCh     = make(chan chan time.Duration)
+	)
+
+	defer ticker.Stop()
+	defer close(pushCh)
+
+	go func() {
+		oldLogs := make([]tmpMsg, 0, h.limit*2)
+		for ch := range pushCh {
+			msgsToPush, msgs = msgs, msgsToPush
+			oldCount, oldDropped := count, dropped
+			count, dropped = 0, 0
+			bufferTime := <-ch
+			close(ch) // signal that more buffering can continue
+
+			copy(oldLogs[len(oldLogs):len(oldLogs)+oldCount], msgsToPush[:oldCount])
+			oldLogs = oldLogs[:len(oldLogs)+oldCount]
+
+			t := time.Now()
+			cutOffIndex := sortAndSplitMsgs(oldLogs, bufferTime)
+			if cutOffIndex == 0 {
+				continue
+			}
+			t1 := time.Since(t)
+
+			strms := h.createPushMessage(oldLogs, cutOffIndex, oldDropped)
+			if cutOffIndex > len(oldLogs) {
+				oldLogs = oldLogs[:0]
+				continue
+			}
+			oldLogs = oldLogs[:copy(oldLogs, oldLogs[cutOffIndex:])]
+			t2 := time.Since(t) - t1
+
+			var b bytes.Buffer
+			_, err := strms.WriteTo(&b)
+			if err != nil {
+				fmt.Printf("Error while marshaling logs for loki %s\n", err)
+				continue
+			}
+			size := b.Len()
+			t3 := time.Since(t) - t2 - t1
+
+			err = h.push(b)
+			if err != nil {
+				fmt.Printf("Error while sending logs to loki %s\n", err)
```

Comment:
I can understand why we're directly printing here, we can't use the logger for obvious reasons. But we should at least print to `stderr`, not `stdout`... 

Even better, in `setupLoggers()` you can make a [new `Logger`](https://godoc.org/github.com/sirupsen/logrus#New) with the same options the original one would have been, and pass that in `LokiFromConfigLine()`. Then you wouldn't even have to mess around with `fmt` here.